src/diff_folded/mod.rs

```diff
@@ -0,0 +1,158 @@
+            counts.first = counts.first * total2 / total1;
+        }
+    }
+    write_stacks(&stack_counts, writer)
+}
+
+/// Produce an output that can be used to generate a differential flame graph from
+/// a before and an after profile.
+///
+/// See [`from_readers`] for the input and output formats.
+pub fn from_files<P1, P2, W>(
+    opt: &Options,
+    filename1: P1,
+    filename2: P2,
+    writer: W,
+) -> io::Result<()>
+where
+    P1: AsRef<Path>,
+    P2: AsRef<Path>,
+    W: Write,
+{
+    let file1 = File::open(filename1)?;
+    let reader1 = io::BufReader::with_capacity(READER_CAPACITY, file1);
+    let file2 = File::open(filename2)?;
+    let reader2 = io::BufReader::with_capacity(READER_CAPACITY, file2);
+    from_readers(opt, reader1, reader2, writer)
+}
+
+// Populate stack_counts based on lines from the reader and returns the sum of the sample counts.
+fn parse_stack_counts<R>(
+    opt: &Options,
+    stack_counts: &mut HashMap<String, Counts>,
+    mut reader: R,
+    is_first: bool,
+) -> io::Result<usize>
+where
+    R: BufRead,
+{
+    let mut total = 0;
+    let mut line = String::new();
+    loop {
+        line.clear();
+
+        if reader.read_line(&mut line)? == 0 {
+            break;
+        }
+
+        if let Some((stack, count)) = parse_line(&line, opt.strip_hex) {
+            let mut counts = stack_counts.entry(stack).or_default();
+            if is_first {
+                counts.first += count;
+            } else {
+                counts.second += count;
+            }
+            total += count;
+        } else {
+            warn!("Unable to parse line: {}", line);
+        }
+    }
+
+    Ok(total)
+}
+
+// Write three-column lines with the folded stack trace and two value columns,
+// one for each profile.
+fn write_stacks<W>(stack_counts: &HashMap<String, Counts>, mut writer: W) -> io::Result<()>
+where
+    W: Write,
+{
+    for (stack, &Counts { first, second }) in stack_counts {
+        writeln!(writer, "{} {} {}", stack, first, second)?;
+    }
+    Ok(())
+}
+
+// Parse stack and sample count from line.
+fn parse_line(line: &str, strip_hex: bool) -> Option<(String, usize)> {
+    let counti = line.rfind(' ')?;
+    let count = &line[(counti + 1)..].trim_end();
+    let count = count.parse::<usize>().ok()?;
+    let mut stack = line[..counti].trim_end().to_string();
+    if strip_hex {
+        stack = strip_hex_address(stack);
+    }
+    Some((stack, count))
+}
+
+// Replace all hex strings like "0x45ef2173" with "0x...".
+fn strip_hex_address(mut stack: String) -> String {
+    let mut start = 0;
+    while let Some(idx) = stack[start..].find("0x") {
+        let ndigits = stack[start + idx + 2..]
+            .chars()
+            .take_while(|c| c.is_digit(16))
+            .count();
+        if ndigits > 0 {
+            stack = format!(
+                "{}0x...{}",
+                &stack[..start + idx],
+                &stack[start + idx + 2 + ndigits..]
+            );
```

Comment:
It might be that it'd be faster to just `push` each character as you iterate past it rather than do multiple `format!` calls. You can even do `String::with_capacity(stack.len())` at the beginning to ensure it won't allocate!